### ISSUE TITLE
No audio output on local devices (eg USB-speakers)

### ISSUE THREAD

**Issue by marioscube** [owntone/owntone-container#7]
I tried owntone from a docker container on debian 12 bookworm xfce.

When I try to play music to airplay devices it works (listening to music now on my homepod mini).

When I try to play music to a locally attached sound device (soundsticks) owntone does not work.

My relevant setup in owntone.conf:

        alsa "hw:1" {
        # Name used in the speaker list. If not set, the card name will be used.
        nickname = "Soundsticks"

        # Mixer channel to use for volume control
        # If not set, PCM will be used if available, otherwise Master
        # mixer = "PCM"

        # Mixer device to use for volume control
        # If not set, the card name will be used
        # mixer_device = ""
        }

Error when enabling the sound device:

`[2024-08-22 09:46:31] [  LOG]   laudio: Cannot get card index for 1[2024-08-22 09:46:31] [  LOG]   laudio: Failed to attach mixer 'hw:1': No such file or directory
[2024-08-22 09:46:31] [  LOG]   laudio: Could not open mixer 'hw:1' ('(null)')
[2024-08-22 09:46:31] [  LOG]      web: JSON api request failed with error code 500 (/api/outputs/0/toggle)
`
![Screenshot 2024-08-22 at 11 46 45](https://github.com/user-attachments/assets/540c0777-1ff7-4d34-8193-e0a8044659a5)

When I use the same config on another computer with owntone installed without docker it works.

Is this a (owntone) docker limitation, a config error or something else?

**Comment by hacketiwack** [comment]
On the principle, containers isolate everything from the host. So this is normal that you can’t access this device from within the container. 
As I never did such configuration myself, I don’t know on the spot how to do it. 
But what you have to look for is to instruct docker or podman to get the device from the host. 
I will investigate in the coming days. And if you find something in between, please comment on that work item.

**Comment by marioscube** [comment]
Thank you for your quick response.

My dirty workaround is:
- install docker [shairport-sync](https://github.com/mikebrady/shairport-sync/blob/master/docker/README.md)
- play audio with owntone to shairport sync (it works)

However i think it would be nice if owntone docker is able to play audio to the local speakers.

**Comment by hacketiwack** [comment]
Great!
The secret ingredient resides in this part of the docker command: `--device /dev/snd`

I need to test that and document it.

**Comment by marioscube** [comment]
I thought so too, however when I tested this it did not work.
Maybe I did something not entirely right?

Please test it for yourself.

**Comment by marioscube** [comment]
@hacketiwack 

The owntone-container does not have a /dev/snd device.

The daapd (owntone) docker from linuxserver.io does have an /dev/snd device.

Adding --device /dev/snd to the startup script works.

Audio plays to my speakers without shairport-sync.

The volume control is not good though, much to loud!

**Comment by hacketiwack** [comment]
@marioscube to be sure to understand.
With the version of linuxserver.io it works when you add the --device /dev/snd to your configuration but the volume is too loud.
Did I understand correctly?
It would be interesting to understand which package is missing so that we can add it as well.

**Comment by marioscube** [comment]
@hacketiwack 

Indeed, if I add —device /dev/snd to the configuration of linuxserver.io daapd to start a docker container local ALSA audio works.

When using the owntone/daapd webinterface I have to slide the volume sliders almost to zero, otherwise my speakers (original usb Harman Kardon soundsticks) are way too loud. That is a problem with stairport-sync as wel. So possibly hardware specific. I need to rest with other speakers.

**Comment by hacketiwack** [comment]
If you can verify which package needs to be added - I guess it is alsalib, plus others - then I could envision to add it.
However, I don't have any way to test it as my motherboard doesn't have audio output at the moment.

Regarding the volume, I guess you have to use the `alsamixer` command.

**Comment by marioscube** [comment]
I will try tomorrow.

No need for an actual working sound card or output on your motherboard.

Just  check inside the container if there is /dev/snd .

Alsamixer does not help enough.

**Comment by marioscube** [comment]
@hacketiwack 
alsa-lib-dev is not the difference (based on a quick compare of build files).

The only relevant difference I see at this moment (without building) is:
    --disable-install_systemd \   # = in owntone / not in daapd

**Comment by marioscube** [comment]
@hacketiwack 

Error on building docker image of owntone with Dockerimage from repository.

I do with dockerfile in directory, from that directory:
  docker build . -t owntone:basic

I get the error:

   => ERROR [runtime 3/4] COPY --chmod=755 /etc/init.d/* /etc/init.d/                                                                                         0.0s
------
 > [runtime 3/4] COPY --chmod=755 /etc/init.d/* /etc/init.d/:
------
Dockerfile:73
--------------------
  71 |     
  72 |     COPY --from=build /tmp/build/ .
  73 | >>> COPY --chmod=755 /etc/init.d/* /etc/init.d/
  74 |     
  75 |     RUN \
--------------------
ERROR: failed to solve: lstat /var/lib/docker/tmp/buildkit-mount172616985/etc/init.d: no such file or directory


**I gave up.**
Linuxserver.io daapd docker also gave an error, so it might not be the Dockerfile of owntone that is not working.

**Comment by marioscube** [comment]
..... interesting ......

I build owntone-docker from this github.

git clone https://github.com/owntone/owntone-container

docker build . -t owntone:test

then:
```
docker run -d \
  --name=owntone \
  --network=host \
  -e UID=1000 \
  -e GID=1000 \
  -v /home/debian/Docker/owntone/config:/etc/owntone \
  -v /home/debian/Docker/owntone/media:/srv/media \
  -v /home/debian/Docker/owntone/cache:/var/cache/owntone \
  **--device /dev/snd**  \
  owntone:test
```

check for /dev/snd :
docker exec -ti owntone sh
  cd /dev/snd
  **existst OK!** (not in owntone from provided image source)


Edit owntone.conf

alsa "hw:1" {
        # Name used in the speaker list. If not set, the card name will be used.
        nickname = "Earplugs"

        # Mixer channel to use for volume control
        # If not set, PCM will be used if available, otherwise Master
#       mixer = "PCM"

        # Mixer device to use for volume control
        # If not set, the card name will be used
#       mixer_device = "PCM"
}

And it works ........ (for now).
After some (a lot of) fiddling with the settings in owntone.conf I have music through my earphones.

@hacketiwack 
I would say build your own owntone docker-image and check if there is a /dev/snd device in the container. Then we can compare if needed.

**Comment by marioscube** [comment]
Aanvullinkje:

Als ik op basis van het hier gegeven buildfile zelf een image van owntone maak mist group 29 in de container. Zelf even toegevoegd in /etc/group file. Lijkt effect te hebben op /dev/snd en dus of en hoe je eigen audio devices kan toevoegen en hoe deze werken.

Moet nog wat verder testen.

**Comment by timbrd** [comment]
> @hacketiwack
> 
> Indeed, if I add —device /dev/snd to the configuration of linuxserver.io daapd to start a docker container local ALSA audio works.
> 
> When using the owntone/daapd webinterface I have to slide the volume sliders almost to zero, otherwise my speakers (original usb Harman Kardon soundsticks) are way too loud. That is a problem with stairport-sync as wel. So possibly hardware specific. I need to test with other speakers.

Same here. I send the audio stream to an av receiver via airplay, but i have to turn the volume nearly down so that it is not too loud. Maybe an alsa issue?
Did you fix the issue with your soundsticks?

**Comment by marioscube** [comment]
@timbrd 

Sending audio to HomePod mini and/or AirPort Express I have no problem with needing to have to turn the volume down.

If I send the audio to my USB Soundsticks I need to turn the volume almost to 0.

I will check again tomorrow.

**Comment by Asik007** [comment]
Hey, I've been banging my head against this for the past two days (not really a Docker person, so AI has been a huge help). Do you still need to work around building the Dockerfile from scratch?

I've tried both a privileged Docker container and `UID:29` / `group_add: audio` (I'll attach the docker-compose file). That said, the logs show:

```
laudio: Failed to probe ALSA card=0 - No such file or directory
```

At this point, I'm not sure if this is something on my end.

[compose.yaml](https://github.com/user-attachments/files/30257785/compose.yaml)

**Comment by marioscube** [comment]
1. it's been a while since I saw this topic, almost 2 years
2. my server with docker/owntone crashed (broken motherboard/CPU/PSU take your pick, I still don't know). I stopped using owntone with docker. 
3. try the docker image from linuxserver.io, that worked (at least 2 years ago)
4. build your own docker image. Do not know if that works. Have not tried.
5. to make owntone work inside a owntone-container docker container you have to manually add group 29 to /etc/group inside the container.

**Comment by Asik007** [comment]
Thank you for the response on such an old thread! Sorry about your late server, mine is an old ThinkPad from highschool so its probably about to croak.

I tried your fix for adding the "owntone" user to the "audio" group **within the container,** and it finally works! (It makes so much sense and I don't know why I never checked that.) 

Thank You!

**Comment by marioscube** [comment]
Take good care of your Tinkpad.
The prices of new hardware are rediculous.

Have fun with owntone!
